crates/ruff_python_ast/src/visitor/preorder.rs

```diff
@@ -0,0 +1,1101 @@
+            subject,
+            cases,
+            range: _range,
+        }) => {
+            visitor.visit_expr(subject);
+            for match_case in cases {
+                visitor.visit_match_case(match_case);
+            }
+        }
+        Stmt::Raise(StmtRaise {
+            exc,
+            cause,
+            range: _range,
+        }) => {
+            if let Some(expr) = exc {
+                visitor.visit_expr(expr);
+            };
+            if let Some(expr) = cause {
+                visitor.visit_expr(expr);
+            };
+        }
+        Stmt::Try(StmtTry {
+            body,
+            handlers,
+            orelse,
+            finalbody,
+            range: _range,
+        })
+        | Stmt::TryStar(StmtTryStar {
+            body,
+            handlers,
+            orelse,
+            finalbody,
+            range: _range,
+        }) => {
+            visitor.visit_body(body);
+            for excepthandler in handlers {
+                visitor.visit_excepthandler(excepthandler);
+            }
+            visitor.visit_body(orelse);
+            visitor.visit_body(finalbody);
+        }
+        Stmt::Assert(StmtAssert {
+            test,
+            msg,
+            range: _range,
+        }) => {
+            visitor.visit_expr(test);
+            if let Some(expr) = msg {
+                visitor.visit_expr(expr);
+            }
+        }
+        Stmt::Import(StmtImport {
+            names,
+            range: _range,
+        }) => {
+            for alias in names {
+                visitor.visit_alias(alias);
+            }
+        }
+        Stmt::ImportFrom(StmtImportFrom {
+            range: _,
+            module: _,
+            names,
+            level: _,
+        }) => {
+            for alias in names {
+                visitor.visit_alias(alias);
+            }
+        }
+
+        Stmt::Pass(_)
+        | Stmt::Break(_)
+        | Stmt::Continue(_)
+        | Stmt::Global(_)
+        | Stmt::Nonlocal(_) => {}
+    }
+}
+
+pub fn walk_expr<'a, V>(visitor: &mut V, expr: &'a Expr)
+where
+    V: PreorderVisitor<'a> + ?Sized,
+{
+    match expr {
+        Expr::BoolOp(ExprBoolOp {
+            op,
+            values,
+            range: _range,
+        }) => {
+            if !values.is_empty() {
+                visitor.visit_expr(&values[0]);
+                visitor.visit_boolop(op);
+                for expr in &values[1..] {
+                    visitor.visit_expr(expr);
+                }
+            }
+        }
+        Expr::NamedExpr(ExprNamedExpr {
+            target,
+            value,
+            range: _range,
+        }) => {
+            visitor.visit_expr(target);
+            visitor.visit_expr(value);
+        }
+        Expr::BinOp(ExprBinOp {
+            left,
+            op,
+            right,
+            range: _range,
+        }) => {
+            visitor.visit_expr(left);
+            visitor.visit_operator(op);
```

Comment:
this is so confusing to me because we have the BinOp node but than there's another node for the actual Op which we visit in-order

Reply:
To make this more confusing. `op` isn't an AST node because it doesn't have a `range`